README.md
ELPickerView: Easily Used Picker View build with Swift 4
======================================

[![Build Status](https://travis-ci.org/Elenionl/ELPickerView-Swift.svg?branch=master)](https://travis-ci.org/Elenionl/ELPickerView-Swift)
[![CocoaPods](https://img.shields.io/cocoapods/v/ELPickerView.svg?style=flat)](https://cocoapods.org/pods/ELPickerView)
[![CocoaPods](https://img.shields.io/cocoapods/l/ELPickerView.svg?style=flat)](https://cocoapods.org/pods/ELPickerView)
[![Swift 4.0](https://img.shields.io/badge/Swift-4.0-blue.svg)](https://img.shields.io/badge/Swift-4.0-blue.svg)
[![iOS 8.0](https://img.shields.io/badge/iOS-8.0-blue.svg)](https://img.shields.io/badge/iOS-8.0-blue.svg)

**:warning: ELPickerView requires Swift Version higher than 4.0.**

## 请点击[中文说明](https://github.com/Elenionl/ELPickerView-Swift/blob/master/README%20IN%20CHINESE.md)


## Screenshots
Easily Used Picker View build with Swift 4


![screenshots_4](https://raw.githubusercontent.com/Elenionl/ELPickerView-Swift/master/screenshots/screenshots_4.gif)
----------------------

## How to Install

### Using [CocoaPods](https://cocoapods.org/pods/ELPickerView)

* Add this line to your ``podfile`` :
``pod 'ELPickerView'``
* Swift 3.* use 3.0.0 pod version
* Swift 4.* use 4.* pod version
* Run `pod install` with Terminal
* Then everything is done!
-----------------------
### Simply add

* Open [Elenionl/ELPickerView-Swift](https://github.com/Elenionl/ELPickerView-Swift) with browser
* Download or Clone Project: ``https://github.com/Elenionl/ELPickerView-Swift.git``
* Copy ``ELCustomPickerView.swift`` file to your project
* Enjoy
-------------

## How to Use

### If you want to show a Picker View in your application. Simply do these two steps:
* init
```Swift
lazy var customPickerView: ELCustomPickerView<String> = {
    return ELCustomPickerView<String>(pickerType: .singleComponent, items: [
        "Row 0"
        , "Row 1"
        , "Row 2"
        , "Row 3"
        , "Row 4"
        , "Row 5"
        ])
}
```
* show
```Swift
override func viewDidLoad() {
        super.viewDidLoad()
        customPickerView.show(viewController: self, animated: true)
}
```
----------------------
### If You Want the Translucent Background to Cover the Navigation Bar and the Tab Background

* Just let the window show Picker View like this:

```Swift
    override func viewDidLoad() {
        super.viewDidLoad()
        customPickerView.show(viewController: nil, animated: true)
    }
```
--------------
### Instead of Delegate, ELPickerView Use Closure to Handle Callback

* There are rich callbacks as follow:
Using seres of 'Set' methods
```Swift
        view.setDidScrollHandler({ [weak self] (view, chosenIndex, chosenItem) -> (shouldHide: Bool, animated: Bool) in
            let hide = false
            let animated = false
            self?.logLabel.text = "Did Scroll. \n<Index: \(chosenIndex)> \n<chosenItem: \(chosenItem)> \n<Hide: \(hide)> \n<Animated: \(animated)>"
            print(self?.logLabel.text ?? "")
            return (hide, animated)
        })
```
set value
```Swift
        customPickerView.leftButtoTapHandler = { [weak self] (view: ELCustomPickerView<String>?, chosenIndex: Int, chosenItem: (key: String, content: String)) in
            let hide = true
            let animated = true
            self?.logLabel.text = "Did Tap Left Button. <Index: \(chosenIndex)> <chosenItem: \(chosenItem)> <Hide: \(hide)> <Animated: \(animated)>"
            print(self?.logLabel.text ?? "")
            return (hide, animated)
        }
```

* And here is the define and meaning of the handler:

```Swift
    /// Triggered when Left Button is tapped.
    //  view: the CustomPickerView
    //  chosenIndex: the current chosen index of row in Picker View
    //  chosenItem: the Item connected with the chosen row
    //  shouldHide: tell the Picker View whether it should be hide  Default value is true
    //  animated: tell the Picker View whether the hide action should have animation  Default value is true
    public var leftButtoTapHandler: ((_ view: ELCustomPickerView?, _ chosenIndex: Int, _ chosenItem: T) -> (shouldHide: Bool, animated: Bool))?
```
--------------------------
### You Can Let Users Pick Between Instance, Struct, Enum, Tuple

* Add whatever kind of items when init:
```Swift
typealias CustomView = ELCustomPickerView<(key: String, content: String)>
...
let view = CustomView(pickerType: .singleComponent, items: [
          ("00", "Row 0")
          , ("02", "Row 1")
          , ("04", "Row 2")
          , ("06", "Row 3")
          , ("09", "Row 4")
          , ("11", "Row 5")
          ])
```

* Give a handler to transform item into a String kind var
```Swift
view.itemConfigHandler = { (key: String, content: String) in
    return content
}
```

* Then Done
--------------------
### View is Easily Customizable

```Swift
        view.blackBackground = true
        view.isTitleBarHidden = false
        view.isTapBackgroundEnabled = true
        view.leftButton.setTitle("LEFT", for: .normal)
        view.rightButton.setTitle("RIGHT", for: .normal)
        view.title.text = "TITLE"
        view.foregroundView.picker.backgroundColor = UIColor.white
        view.foregroundView.bottomDivider.isHidden = true
```
-----------------
## Settings and Handlers Available

```Swift
// MARK: - Settings

/// Type of Picker View
public let pickerType: ELCustomPickerViewType

/// Items used to config Picker View rows  Default value is []
public var items: [T]

/// Background of the screen  Default value is true
public var blackBackground: Bool

/// Set Title Bar hidden or not  Default value is false
public var isTitleBarHidden = false

/// Set Taping Background to hide Picker View enabled or not  Default value is true
public var isTapBackgroundEnabled = true

/// Left Button of the Title Bar, shortcut to foregroundView.leftButton
public lazy var leftButton: UIButton

/// Right Button of the Title Bar, shortcut to foregroundView.rightButton
public lazy var rightButton: UIButton

/// Title of the Title Bar, shortcut to foregroundView.title
public lazy var title: UILabel


// MARK: - Handler

/// Function used to transform Item into String. If the Item is String kind, itemConfigHandler is not necessory to be set.
public var itemConfigHandler: ((T) -> String)?

/// Triggered when Left Button is tapped.
//  view: the CustomPickerView
//  chosenIndex: the current chosen index of row in Picker View
//  chosenItem: the Item connected with the chosen row
//  shouldHide: tell the Picker View whether it should be hide  Default value is true
//  animated: tell the Picker View whether the hide action should have animation  Default value is true
public var leftButtoTapHandler: ((_ view: ELCustomPickerView?, _ chosenIndex: Int, _ chosenItem: T) -> (shouldHide: Bool, animated: Bool))?

/// Triggered when Right Button is tapped.
//  view: the CustomPickerView
//  chosenIndex: the current chosen index of row in Picker View
//  chosenItem: the Item connected with the chosen row
//  shouldHide: tell the Picker View whether it should be hide  Default value is true
//  animated: tell the Picker View whether the hide action should have animation  Default value is true
public var rightButtoTapHandler: ((_ view: ELCustomPickerView?, _ chosenIndex: Int, _ chosenItem: T) -> (shouldHide: Bool, animated: Bool))?

/// Triggered when user picked one row in Picker View.
//  view: the CustomPickerView
//  chosenIndex: the current chosen index of row in Picker View
//  chosenItem: the Item connected with the chosen row
//  shouldHide: tell the Picker View whether it should be hide  Default value is false
//  animated: tell the Picker View whether the hide action should have animation   Default value is false
public var didScrollHandler: ((_ view: ELCustomPickerView?, _ chosenIndex: Int, _ chosenItem: T) -> (shouldHide: Bool, animated: Bool))?

/// Triggered when Picker View will show
public var willShowHandler: ((_ view: ELCustomPickerView?) -> Void)?

/// Triggered when Picker View did show
public var didShowHandler: ((_ view: ELCustomPickerView?) -> Void)?

/// Triggered when Picker View will hide
public var willHideHandler: ((_ view: ELCustomPickerView?) -> Void)?

/// Triggered when Picker View did hide
public var didHideHandler: ((_ view: ELCustomPickerView?) -> Void)?

// MARK: - Views

/// The bottom view containing Title Bar and Picker
public lazy var foregroundView: ELPickerForegroundView
```
------------
## Requirements

* Xcode 8.0
* Swift 4.0
* Using ARC
* iOS 8.0
--------------

## TODO
* ❌ Picker View with more than one component
* ❌ Picker View of date and time
## Author

Hanping Xu ([Elenionl](https://github.com/Elenionl)), stellanxu@gmail.com


--------------------------
## License

ELPickerView is available under the MIT license, see the LICENSE file for more information.   


CHANGELOG.md
## 0.1.4
* Fix background tap bug.
* Make handlers more easy to use.
## 3.0.0
* Last Version for Swift 3
## 4.0.0
* Support Swift 4
## 4.0.1
* update readme

ELPickerView/Demo/ViewController.swift
/*
**The MIT License**
**Copyright © 2017 Hanping Xu**
**All rights reserved.**

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

import UIKit

typealias CustomView = ELCustomPickerView<(key: String, content: String)>

class ViewController: UIViewController {
    @IBOutlet weak var logLabel: UILabel!
    @IBOutlet weak var showStatus: UILabel!

    @IBAction func showByWindow(_ sender: UIButton) {
        customPickerView.title.text = "Show by Window"
        customPickerView.isTapBackgroundEnabled = true
        customPickerView.blackBackground = true
        customPickerView.isTitleBarHidden = false
        customPickerView.foregroundView.bottomDivider.isHidden = false
        customPickerView.foregroundView.bottomDivider.isHidden = false
        customPickerView.show(viewController: nil, animated: true)
    }
    @IBAction func showByViewController(_ sender: UIButton) {
        customPickerView.title.text = "Show by ViewController"
        customPickerView.isTapBackgroundEnabled = true
        customPickerView.blackBackground = true
        customPickerView.isTitleBarHidden = false
        customPickerView.foregroundView.bottomDivider.isHidden = false
        customPickerView.foregroundView.bottomDivider.isHidden = false
        customPickerView.show(viewController: self, animated: true)
    }
    @IBAction func showWithWhiteBackground(_ sender: UIButton) {
        customPickerView.title.text = "Show With White Background"
        customPickerView.isTapBackgroundEnabled = true
        customPickerView.blackBackground = false
        customPickerView.isTitleBarHidden = false
        customPickerView.foregroundView.bottomDivider.isHidden = false
        customPickerView.foregroundView.bottomDivider.isHidden = false
        customPickerView.show(viewController: self, animated: true)
    }
    @IBAction func showWithoutTitleBar(_ sender: UIButton) {
        customPickerView.title.text = "Show With White Background"
        customPickerView.isTapBackgroundEnabled = true
        customPickerView.blackBackground = true
        customPickerView.isTitleBarHidden = true
        customPickerView.foregroundView.bottomDivider.isHidden = false
        customPickerView.foregroundView.bottomDivider.isHidden = false
        customPickerView.show(viewController: self, animated: true)
    }
    @IBAction func showWithNoDivider(_ sender: UIButton) {
        customPickerView.title.text = "Show With No Divider & Tap Background Disabled"
        customPickerView.isTapBackgroundEnabled = false
        customPickerView.blackBackground = true
        customPickerView.isTitleBarHidden = false
        customPickerView.foregroundView.bottomDivider.isHidden = true
        customPickerView.foregroundView.bottomDivider.isHidden = true
        customPickerView.show(viewController: self, animated: true)
    }

    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
    }
    
    lazy var customPickerView: CustomView = {
        let view = CustomView(pickerType: .singleComponent, items: [
            ("00", "Row 0")
            , ("02", "Row 1")
            , ("04", "Row 2")
            , ("06", "Row 3")
            , ("09", "Row 4")
            , ("11", "Row 5")
            ])
        // MARK: - Settings
        view.blackBackground = true
        view.isTitleBarHidden = false
        view.isTapBackgroundEnabled = true
        view.leftButton.setTitle("LEFT", for: .normal)
        view.rightButton.setTitle("RIGHT", for: .normal)
        view.title.text = "TITLE"
        view.foregroundView.picker.backgroundColor = UIColor.white
        view.foregroundView.bottomDivider.isHidden = true
        // MARK: - Handler
        view.setItemConfigHandler({ (item) -> String in
            return item.content
        })
        view.setLeftButtoTapHandler({ [weak self] (view, chosenIndex, chosenItem) -> (shouldHide: Bool, animated: Bool) in
            let hide = true
            let animated = true
            self?.logLabel.text = "Did Tap Left Button.\n <Index: \(chosenIndex)> \n<chosenItem: \(chosenItem)> \n<Hide: \(hide)> \n<Animated: \(animated)>"
            print(self?.logLabel.text ?? "")
            return (hide, animated)
        })
        view.setRightButtoTapHandler({ [weak self] (view, chosenIndex, chosenItem) -> (shouldHide: Bool, animated: Bool) in
            let hide = true
            let animated = true
            self?.logLabel.text = "Did Tap Right Button. \n<Index: \(chosenIndex)> \n<chosenItem: \(chosenItem)> \n<Hide: \(hide)> \n<Animated: \(animated)>"
            print(self?.logLabel.text ?? "")
            return (hide, animated)
        })
        view.setDidScrollHandler({ [weak self] (view, chosenIndex, chosenItem) -> (shouldHide: Bool, animated: Bool) in
            let hide = false
            let animated = false
            self?.logLabel.text = "Did Scroll. \n<Index: \(chosenIndex)> \n<chosenItem: \(chosenItem)> \n<Hide: \(hide)> \n<Animated: \(animated)>"
            print(self?.logLabel.text ?? "")
            return (hide, animated)
        })
        view.setWillShowHandler({ [weak self] (view) in
            self?.showStatus.text = "View Will Show"
            print(self?.showStatus.text ?? "")
        })
        view.setDidShowHandler({ [weak self] (view) in
            self?.showStatus.text = "View Did Show"
            print(self?.showStatus.text ?? "")
        })
        view.setWillHideHandler({ [weak self] (view) in
            self?.showStatus.text = "View Will Hide"
            print(self?.showStatus.text ?? "")
        })
        view.setDidHideHandler({ [weak self] (view) in
            self?.showStatus.text = "View Did Hide"
            print(self?.showStatus.text ?? "")
        })
        return view
    }()
}

